### PR TITLE
Fix issue #395 and add regression tests

### LINKED ISSUE #298: Returning a tuple from a pure function causes Prusti to crash
The following program:

```rust
use prusti_contracts::*;
#[pure]
fn dummy() { }
fn main() {}
```

gives a stacktrace:

```plain
hread 'rustc' panicked at 'not implemented: Tuple([])', prusti-viper/src/encoder/encoder.rs:1071:22
stack backtrace:                                                  
   0: rust_begin_unwind
             at /rustc/98d66340d6e63eda115afc8b0da1d87965881936/library/std/src/panicking.rs:495:5
   1: std::panicking::begin_panic_fmt
             at /rustc/98d66340d6e63eda115afc8b0da1d87965881936/library/std/src/panicking.rs:437:5
   2: prusti_viper::encoder::encoder::Encoder::encode_const_expr
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/encoder.rs:1071:22
   3: prusti_viper::encoder::mir_encoder::MirEncoder::encode_operand_expr
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/mir_encoder.rs:383:19
   4: <prusti_viper::encoder::pure_function_encoder::PureFunctionBackwardInterpreter as prusti_viper::encoder::mir_interpreter::BackwardMirInterpreter>::apply_statement
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/pure_function_encoder.rs:931:48
   5: prusti_viper::encoder::mir_interpreter::run_backward_interpretation
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/mir_interpreter.rs:74:13
   6: prusti_viper::encoder::pure_function_encoder::PureFunctionEncoder::encode_function
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/pure_function_encoder.rs:81:25
   7: prusti_viper::encoder::encoder::Encoder::encode_pure_function_def
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/encoder.rs:1223:37
   8: prusti_viper::encoder::encoder::Encoder::process_encoding_queue
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/encoder/encoder.rs:1466:17
   9: prusti_viper::verifier::Verifier::verify
             at /home/data/eth/rust/prusti-dev/prusti-viper/src/verifier.rs:243:9
  10: prusti_driver::verifier::verify
             at /home/data/eth/rust/prusti-dev/prusti/src/verifier.rs:48:39
  11: <prusti_driver::callbacks::PrustiCompilerCalls as rustc_driver::Callbacks>::after_analysis::{{closure}}
             at /home/data/eth/rust/prusti-dev/prusti/src/callbacks.rs:82:17
  12: rustc_interface::passes::QueryContext::enter::{{closure}}
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_interface/src/passes.rs:729:42
  13: rustc_middle::ty::context::tls::enter_context::{{closure}}
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_middle/src/ty/context.rs:1719:50
  14: rustc_middle::ty::context::tls::set_tlv
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_middle/src/ty/context.rs:1703:9
  15: rustc_middle::ty::context::tls::enter_context
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_middle/src/ty/context.rs:1719:9
  16: rustc_interface::passes::QueryContext::enter
             at /home/vakaras/.rustup/toolchains/nightly-2020-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/compiler/rustc_interface/src/passes.rs:729:9
  17: <prusti_driver::callbacks::PrustiCompilerCalls as rustc_driver::Callbacks>::after_analysis
             at /home/data/eth/rust/prusti-dev/prusti/src/callbacks.rs:48:9
  18: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  19: rustc_span::with_source_map
  20: rustc_interface::interface::create_compiler_and_run
  21: rustc_span::with_session_globals
```

Since `()` is `Copy`, this code should be legal.

cc @cmatheja

### LINKED ISSUE #395: ICE: returning a tuple, or tuple struct, from a function
The following two code snippets each cause an internal compiler error in prusti commit 6b551ab:

```rust
#[pure]
fn f() -> (u8, u8) {
    (0, 0)
}
```

```rust
#[derive(Copy, Clone)]
struct X(u8, u8);

#[pure]
fn f() -> X {
    X(0, 0)
}
```

The error originates in the `expect` call in https://github.com/viperproject/prusti-dev/blob/master/prusti-viper/src/encoder/pure_function_encoder.rs#L257-L259:

```
failed generation of folding/unfolding
```

For reference, this code compiles successfully (I didn't know it was valid Rust before today):

```rust
#[pure]
fn f() -> (u8) {
    (0)
}
```

### PR DESCRIPTION
**Pull request by fpoli**
Closes #395.
Closes #298.

### CODE REVIEW

File: prusti-viper/src/encoder/mir/types/encoder.rs [hunk 1]
```diff
@@ -259,12 +260,7 @@ impl<'p, 'v, 'r: 'v, 'tcx: 'v> TypeEncoder<'p, 'v, 'tcx> {
                 Some(bounds)
             }
             ty::TyKind::Char => Some((0.into(), std::char::MAX.into())),
-            ty::TyKind::Bool | ty::TyKind::Ref(_, _, _) => None,
-            ty::TyKind::Adt(_, _) => {
-                assert!(self.encoder.env().type_is_copy(self.ty));
```
Comment (fpoli) [inline prusti-viper/src/encoder/mir/types/encoder.rs:264]: @vakaras Do you remember why you added this assert in 70ad6e91f023f12d674c71b6d1d092b1c1643d63?

Comment (vakaras) [reply prusti-viper/src/encoder/mir/types/encoder.rs:264]: This code is called only for pure functions, which require all types to be `Copy`.

Comment (vakaras) [reply prusti-viper/src/encoder/mir/types/encoder.rs:264]: In other words: a sanity check.

Comment (fpoli) [reply prusti-viper/src/encoder/mir/types/encoder.rs:264]: https://github.com/viperproject/prusti-dev/pull/736 moves the check to the call sites.

**Pull request merged**